### PR TITLE
set proper image dimensions

### PR DESCRIPTION
**Pull request by quarterto**
[
screenshot 2016-02-25 12 13 59
](http://ft-google-amp-staging-pr-41.herokuapp.com/content/937c1f92-91e0-11e3-adde-00144feab7de)

### CODE REVIEW

File: server/lib/body-transform.js [hunk 1]
```diff
@@ -28,11 +14,13 @@ module.exports = function run(body, flags) {
...
-		.with(externalImages)
-		.with(trimmedLinks)
-		.get();
+	const transformed$ = [
+		externalImages,
+		trimmedLinks,
+	].reduce(
+		(promise$, transform) => promise$.then($ => transform($, flags)),
+		Promise.resolve(cheerio.load(body, {decodeEntities: false}))
+	);
```
Comment (quarterto) [inline server/lib/body-transform.js]: I really do love promises

File: server/lib/external-images.js [hunk 2]
```diff
@@ -1,31 +1,43 @@
...
-			if(!$el.attr('height')) {
-				$el.attr('width', '700px');
-				$el.attr('height', '400px');
-			}
+			const metaUrl = fullSrcEncoded.replace('raw', 'metadata');
+			return fetch(metaUrl)
+				.then(response => response.json())
+				.then(
+					meta => Object.assign(meta, {ratio: meta.height / meta.width}),
+					(e) => (console.error(e), {width: 600, ratio: 4 / 7}) // discard error and use fallback dimensions
```
Comment (georgecrawford) [inline server/lib/external-images.js]: Ooh, I like the `=> (console.log(), {return: 'this'})` pattern :)

Comment (quarterto) [inline server/lib/external-images.js]: Oops, that shouldn't have been left in. Maybe I'll add Raven logging here.

File: server/lib/external-images.js [hunk 3]
```diff
@@ -1,31 +1,43 @@
...
+						$el.attr('layout', 'fixed');
+					}
+				});
 		}
 
 		return $el;
-	});
-
-	return $;
+	}).toArray()).then(() => $);
```
Comment (georgecrawford) [inline server/lib/external-images.js]: Can you explain the `toArray()`?

**Comment by georgecrawford** [comment]
LGTM

File: server/lib/external-images.js (hunk 3 above)
Comment (quarterto) [inline server/lib/external-images.js]: Cheerio objects are array-like, not real arrays (they're akin to a `NodeList`), but `Promise.all` expects a real array or iterable.

Comment (georgecrawford) [inline server/lib/external-images.js]: gotcha

**Pull request merged**